File: NationalSecurityCommunity/app/src/main/java/com/national/security/community/base/BaseFragment.java
package com.national.security.community.base;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.national.security.community.App;
import com.national.security.community.injection.component.DaggerFragmentComponent;
import com.national.security.community.injection.component.FragmentComponent;
import com.national.security.community.injection.module.FragmentModule;
import com.national.security.community.mvp.IPresenter;
import com.national.security.community.mvp.IView;

import javax.inject.Inject;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 @ description:  基类Fragment
 @ author:  ljn
 @ time:  2018/1/2 
 */
public abstract class BaseFragment<T extends IPresenter> extends Fragment implements IView {

    @Inject
    protected T mPresenter;
    protected View mView;
    protected Activity mActivity;
    protected Context mContext;
    private Unbinder mUnBinder;

    @Override
    public void onAttach(Context context) {
        mActivity = (Activity) context;
        mContext = context;
        super.onAttach(context);
    }

    protected FragmentComponent getFragmentComponent() {
        return DaggerFragmentComponent.builder()
                .appComponent(App.getAppComponent())
                .fragmentModule(getFragmentModule())
                .build();
    }

    protected FragmentModule getFragmentModule() {
        return new FragmentModule(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mView = inflater.inflate(getLayoutId(), null);
        initInject();
        return mView;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initEventAndData();
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (mPresenter != null) mPresenter.attachView(this);
        mUnBinder = ButterKnife.bind(this, view);
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mUnBinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mPresenter != null) {
            mPresenter.detachView();
        }
    }


    protected abstract void initInject();

    protected abstract int getLayoutId();

    protected abstract void initEventAndData();
}

File: app/src/main/java/com/jmd_news_kotlin/base/BaseFragment.kt
package com.jmd_news_kotlin.base

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.jmd_news_kotlin.App
import com.jmd_news_kotlin.di.component.DaggerFragmentComponent
import com.jmd_news_kotlin.di.component.FragmentComponent
import com.jmd_news_kotlin.di.module.FragmentModule
import com.jmd_news_kotlin.mvp.IPresenter
import com.jmd_news_kotlin.mvp.IView
import me.yokeyword.fragmentation.SupportFragment
import javax.inject.Inject

abstract class BaseFragment<T : IPresenter> : SupportFragment(), IView {

    @Inject
    @JvmField
    protected var mPresenter: T? = null
    protected var mActivity: Activity? = null
    protected var mContext: Context? = null
    protected var isInited = false
    protected var mView: View? = null
    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mActivity = context as Activity
        mContext = context
    }

    protected fun getFragmentComponent(): FragmentComponent {
        return DaggerFragmentComponent.builder()
                .appComponent(App.appComponent)
                .fragmentModule(FragmentModule(this))
                .build()
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mView = inflater?.inflate(getLayoutId(), null)
        initInject()
        return mView
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (savedInstanceState == null) {
            if (!isHidden) {
                isInited = true
                initEventAndData()
            }
        } else {
            if (isSupportVisible) {
                isInited = true
                initEventAndData()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter?.detachView()
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!isInited && !hidden) {
            isInited = true
            initEventAndData()
        }
    }

    protected abstract fun initInject()
    protected abstract fun getLayoutId(): Int
    protected abstract fun initEventAndData()
}